### PR TITLE
Update the hash algorithm used to perform a one-way hash of the token.

### PR DESCRIPTION
**Pull request by andralungu**

### CODE REVIEW

**Review by atrifan — CHANGES_REQUESTED** [review]
https://github.com/openresty/lua-resty-string - please see how the resty.aes and resty.sha256 works also all you need to do is maybe expose that under a factory class.

**Comment by andralungu** [comment]
@atrifan please have another look.

File: src/lua/api-gateway/util/hasher.lua [hunk 1]
```diff
@@ -0,0 +1,39 @@
...
+
+local resty_sha256 = require "resty.sha256"
+local str = require "resty.string"
+
+---
+-- Encrypts the plain_text with a specific salt using the SHA256 algorithm.
+-- @param plain_text The Text to encode
+-- @return - the encrypted text
+--
+function _hash(plain_text)
```
Comment (corinapurcarea) [inline src/lua/api-gateway/util/hasher.lua]: make it local if you're using return {...}

File: src/lua/api-gateway/util/hasher.lua [hunk 2]
```diff
@@ -0,0 +1,39 @@
...
+    local sha256 = resty_sha256:new()
+    sha256:update(plain_text)
+    local digest = sha256:final()
+    ngx.log(ngx.DEBUG, "Hashed ", plain_text, " - ", digest)
```
Comment (corinapurcarea) [inline src/lua/api-gateway/util/hasher.lua]: Is is helpful to log the binary digest?

**Review by corinapurcarea — CHANGES_REQUESTED** [review]

File: src/lua/api-gateway/util/hasher.lua (hunk 2 above)
Comment (andralungu) [reply src/lua/api-gateway/util/hasher.lua]: Will remove. Needed it for debugging purposes.

**Comment by andralungu** [comment]
@corinapurcarea, please have another look

File: src/lua/api-gateway/util/hasher.lua [hunk 3]
```diff
@@ -0,0 +1,38 @@
...
+--   AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+--   LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING
+--   FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER
+--   DEALINGS IN THE SOFTWARE.
+
+local resty_sha256 = require "resty.sha256"
+local str = require "resty.string"
+
+---
+-- Encrypts the plain_text with a specific salt using the SHA256 algorithm.
```
Comment (corinapurcarea) [inline src/lua/api-gateway/util/hasher.lua]: update the comment

**Review by corinapurcarea — CHANGES_REQUESTED** [review]

**Review by corinapurcarea — CHANGES_REQUESTED** [review]
Merge master in this branch to have the support for adding unit tests for hasher and generate documentation

**Comment by corinapurcarea** [comment]
Don't forget to update the rockspec before the release

**Comment by andralungu** [comment]
Rebased against master.

**Review by atrifan — APPROVED** [review]

**Comment by atrifan** [comment]
@corinapurcarea @vdatcu looks good now please check

**Review by corinapurcarea — APPROVED** [review]
👍

File: test/perl/api-gateway/validation/oauth2/oauthTokenValidator.t [hunk 4]
```diff
@@ -130,21 +132,26 @@ env REDIS_PASS_OAUTH;
             set $authtoken $http_authorization;
             set_if_empty $authtoken $arg_user_token;
             set_by_lua $authtoken 'return ngx.re.gsub(ngx.arg[1], "bearer ", "","ijo") ' $authtoken;
-            set_md5 $authtoken_hash $authtoken;
 
-            set $key 'cachedoauth:$authtoken_hash';
             content_by_lua '
+                local hasher = require "api-gateway.util.hasher"
+                local oauthTokenHash = ngx.var.authtoken_hash
```
Comment (corinapurcarea) [inline test/perl/api-gateway/validation/oauth2/oauthTokenValidator.t:138]: Is this line necessary?

File: test/perl/api-gateway/validation/oauth2/oauthTokenValidator.t [hunk 5]
```diff
@@ -153,14 +160,19 @@ env REDIS_PASS_OAUTH;
             set $authtoken $http_authorization;
             set_if_empty $authtoken $arg_user_token;
             set_by_lua $authtoken 'return ngx.re.gsub(ngx.arg[1], "bearer ", "","ijo") ' $authtoken;
-            set_md5 $authtoken_hash $authtoken;
-            # set $redis_ttl_cmd 'TTL cachedoauth:$authtoken_hash';
-            set $key 'cachedoauth:$authtoken_hash';
+
             content_by_lua '
+                local hasher = require "api-gateway.util.hasher"
+                local oauthTokenHash = ngx.var.authtoken_hash
```
Comment (corinapurcarea) [inline test/perl/api-gateway/validation/oauth2/oauthTokenValidator.t:166]: Is this line necessary?

File: test/perl/api-gateway/validation/oauth2/userProfileValidator.t [hunk 6]
```diff
@@ -91,32 +93,44 @@ env REDIS_PASS_OAUTH;
             set $authtoken $http_authorization;
             set_if_empty $authtoken $arg_user_token;
             set_by_lua $authtoken 'return ngx.re.gsub(ngx.arg[1], "bearer ", "","ijo") ' $authtoken;
-            set_md5 $authtoken_hash $authtoken;
-            set $key 'cachedoauth:$authtoken_hash';
+
             content_by_lua '
+                local hasher = require "api-gateway.util.hasher"
+                local oauthTokenHash = ngx.var.authtoken_hash
```
Comment (corinapurcarea) [inline test/perl/api-gateway/validation/oauth2/userProfileValidator.t:99]: Is this line necessary?

**Review by corinapurcarea — CHANGES_REQUESTED** [review]

File: src/lua/api-gateway/util/hasher.lua [hunk 7]
```diff
@@ -0,0 +1,49 @@
...
+-- @return - the encrypted text
+--
+local function _hash(plain_text)
+    local algorithm = ngx.var.hashing_algorithm
+    if (algorithm == nil or algorithm == '') then
+        ngx.log(ngx.INFO, "No hashing algorithm has been passed. Defaulting to SHA256")
+        algorithm = "sha256"
+    end
+    if (algorithm ~= "sha256" and algorithm ~= "sha224" and algorithm ~= "sha512" and algorithm ~= "sha384") then
+        ngx.log(ngx.INFO, "The hashing algorithm passed is invalid. Defaulting to SHA256")
```
Comment (vdatcu) [inline src/lua/api-gateway/util/hasher.lua]: You should add a local table containing all the valid algorithms and then write a for loop(extracted in a function `contains` or a similar name) too see if the provided algorithm is in the list. It's easier to see and maintain. Consider 12 new algorithms are to be added.

File: test/unit-tests/api-gateway/util/hasherTest.lua [hunk 8]
```diff
@@ -0,0 +1,70 @@
...
+
+    ngx.var.hashing_algorithm = "sha224"
+    result = classUnderTest.hash("test")
+    assertEquals(result, "sha224")
+
+
+    ngx.var.hashing_algorithm = "sha384"
+    result = classUnderTest.hash("test")
+    assertEquals(result, "sha384")
+end)
```
Comment (vdatcu) [inline test/unit-tests/api-gateway/util/hasherTest.lua:70]: :+1:

**Review by vdatcu — APPROVED** [review]

File: test/perl/api-gateway/validation/oauth2/oauthTokenValidator.t (hunk 4 above)
Comment (andralungu) [reply test/perl/api-gateway/validation/oauth2/oauthTokenValidator.t:138]: It is used to form the cached key for Redis 
```key = "cachedoauth:" .. oauthTokenHash```

File: test/perl/api-gateway/validation/oauth2/oauthTokenValidator.t (hunk 5 above)
Comment (andralungu) [reply test/perl/api-gateway/validation/oauth2/oauthTokenValidator.t:166]: Same comment as above.

File: test/perl/api-gateway/validation/oauth2/userProfileValidator.t (hunk 6 above)
Comment (andralungu) [reply test/perl/api-gateway/validation/oauth2/userProfileValidator.t:99]: Same comment as above.

**Comment by andralungu** [comment]
@vdatcu please have another look.

**Review by vdatcu — APPROVED** [review]

File: test/perl/api-gateway/validation/oauth2/oauthTokenValidator.t (hunk 4 above)
Comment (corinapurcarea) [reply test/perl/api-gateway/validation/oauth2/oauthTokenValidator.t:138]: Actually, it's not needed; you have 2 assignments for local variable oauthTokenHash :
                local oauthTokenHash = ngx.var.authtoken_hash
                ...
                oauthTokenHash = hasher.hash(ngx.var.authtoken)

**Review by atrifan — APPROVED** [review]

**Comment by atrifan** [comment]
@corinapurcarea ?

**Pull request merged**